lib/storage/providers/SQLiteStorage.js

```diff
@@ -0,0 +1,106 @@
+
+const DB_NAME = 'Expensify-new-db';
+QuickSQLite.open(DB_NAME);
+
+QuickSQLite.execute(DB_NAME, 'CREATE TABLE IF NOT EXISTS magic_map (record_key TEXT NOT NULL PRIMARY KEY , valueJSON JSON NOT NULL) WITHOUT ROWID;');
+QuickSQLite.execute(DB_NAME, 'PRAGMA CACHE_SIZE=-20000;');
+QuickSQLite.execute(DB_NAME, 'PRAGMA SYNCHRONOUS=NORMAL;');
+QuickSQLite.execute(DB_NAME, 'PRAGMA journal_mode=WAL;');
+
+const provider = {
+    /**
+      * Get the value of a given key or return `null` if it's not available in storage
+      * @param {String} key
+      * @return {Promise<*>}
+      */
+    getItem(key) {
+        return QuickSQLite.executeAsync(DB_NAME, 'SELECT record_key, valueJSON from magic_map where record_key=?;', [key]).then(({rows}) => {
+            // eslint-disable-next-line no-underscore-dangle
+            const res = rows._array[0];
+            return JSON.parse(res.valueJSON);
+        });
+    },
+
+    /**
+      * Get multiple key-value pairs for the give array of keys in a batch
+      * @param {String[]} keys
+      * @return {Promise<Array<[key, value]>>}
+      */
+    multiGet(keys) {
+        return QuickSQLite.executeAsync(DB_NAME, `SELECT record_key, valueJSON from magic_map where record_key IN (${new Array(keys.length).fill('?').join(',')});`, keys)
+            .then(({rows}) => {
+                // eslint-disable-next-line no-underscore-dangle
+                const res = _.map(rows._array, row => [row.record_key, JSON.parse(row.valueJSON)]);
+                return res;
+            });
+    },
+
+    /**
+      * Sets the value for a given key. The only requirement is that the value should be serializable to JSON string
+      * @param {String} key
+      * @param {*} value
+      * @return {Promise<void>}
+      */
+    setItem(key, value) {
+        return QuickSQLite.executeAsync(DB_NAME, 'REPLACE into magic_map (record_key, valueJSON) VALUES (?, ?);', [key, JSON.stringify(value)]);
+    },
+
+    /**
+      * Stores multiple key-value pairs in a batch
+      * @param {Array<[key, value]>} pairs
+      * @return {Promise<void>}
+      */
+    multiSet(pairs) {
+        const stringifiedArgs = _.map(pairs, ele => [
+            ele[0],
+            JSON.stringify(ele[1]),
+        ]);
+        return QuickSQLite.executeBatchAsync(DB_NAME, [['REPLACE into magic_map (record_key, valueJSON) VALUES (?, json(?));', stringifiedArgs]]);
+    },
+
+    /**
+      * Multiple merging of existing and new values in a batch
+      * @param {Array<[key, value]>} pairs
+      * @return {Promise<void>}
+      */
+    multiMerge(pairs) {
+        return QuickSQLite.executeBatchAsync(DB_NAME,
+            [['INSERT into magic_map (record_key, valueJSON) VALUES (?, json(?)) ON CONFLICT DO UPDATE SET valueJSON = json_patch(valueJSON, json(?));',
+                _.map(pairs, (ele) => {
+                    const str = JSON.stringify(ele[1]);
+                    return [ele[0], str, str];
+                }),
+            ]]);
```

Comment:
> Curious about the second str in this line. Is it needed? If so, maybe a quick comment about why would be nice to have.

It's because there's three `?` in the query, and you need to [bind params](https://www.sqlite.org/c3ref/bind_blob.html) for each. I think we could/should used named bindings like `:param` instead.

Few questions/comments:

- Can you include a note about why we're using `ON CONFLICT DO UPDATE`? It took me a bit to realize that the reason we're not using `INSERT OR REPLACE INTO` here is because if there's already a json value for that key, then we want to use `JSON_PATCH` to deep-merge the new json value into the existing one.
- Can we make sure we are consistent about using `ALL CAPS` for SQLite keywords?
- Can we dig into the QuickSQLite internals here and make sure that `executeBatchAsync` wraps all the separate queries it runs into a single transaction with `BEGIN TRANSACTION` and `COMMIT`?

So overall my suggestion might end up looking like this:

```js
multiMerge(pairs) {
    // Note: We use `ON CONFLICT DO UPDATE` here instead of `INSERT OR REPLACE INTO`
    // so the new JSON value is merged into the old one if there's an existing value
    const query = `
        INSERT OR REPLACE INTO magic_map (record_key, valueJSON)
        VALUES (:key, JSON(:value))
        ON CONFLICT DO UPDATE
        SET valueJSON = JSON_PATCH(valueJSON, JSON(:value));
    `;
    const queryArguments = _.map(timestamp, ([key, value]) => ({key, value}));
    return QuickSQLite.executeBatchAsync(DB_NAME, [query, queryArguments]);
}
```

That also makes me wonder if the `pairs` argument should be `Array<{key, value}>` instead of `Array<[key, value]>`, skipping the need for this `_.map` here

Reply:
Looks like we are good with the `BEGIN TRANSACTION` and `COMMIT` stuff: https://github.com/ospfranco/react-native-quick-sqlite/blob/cea63588d702529c0655408754621cc110275618/cpp/sqlbatchexecutor.cpp#L59-L75

Reply:
> That also makes me wonder if the `pairs` argument should be `Array<{key, value}>` instead of `Array<[key, value]>`, skipping the need for this `_.map` here

This is dictated by the Provider interface - Onyx expects to talk to a storage provider that accepts the current structure of `pairs`, if we change that, we'll have to update a bunch of other places